### ISSUE TITLE
Bug: incorrect file enumeration with legacy config in Node.js 23

### ISSUE THREAD

**Issue by fasttime** [eslint/eslint#19031]
### Environment

Node version: v23.0.0
npm version: v10.9.0
Local ESLint version: v9.12.0 (Currently used)
Global ESLint version: Not found
Operating System: win32 10.0.26100

### What parser are you using?

Default (Espree)

### What did you do?

I installed Node.js 23 and run the tests locally with `npm test`.

### What did you expect to happen?

No errors.

### What actually happened?

I'm getting a reproducible error:

```console
  1) FileEnumerator
       'iterateFiles(patterns)' method should iterate files and configs.
         with 'tests/fixtures/glob-utils' files
           listFilesToProcess()
             should return unignored files from folders unignored in .eslintignore:
     Error: No files matching '...\eslint\tests\fixtures\glob-util\unignored\**\*.js' were found.
      at FileEnumerator.iterateFiles (lib\cli-engine\file-enumerator.js:320:27)
      at iterateFiles.next (<anonymous>)
      at Function.from (<anonymous>)
      at listFiles (tests\lib\cli-engine\file-enumerator.js:274:30)
      at Context.<anonymous> (tests\lib\cli-engine\file-enumerator.js:546:36)
      at process.processImmediate (node:internal/timers:511:21)
```

### Link to Minimal Reproducible Example

https://github.com/eslint/eslint/blob/v9.12.0/tests/lib/cli-engine/file-enumerator.js#L542-L552

### Participation

- [x] I am willing to submit a pull request for this issue.

### Additional comments

Node.js 23 fixes some inconsistencies in the way [`node:path`](https://nodejs.org/api/path.html) functions were working in previous versions (see PR https://github.com/nodejs/node/pull/54224).

As a result of this change, some path-handling functions can now produce different results depending on the version of Node.js. For example:

```
import { relative } from "node:path";

const basePath = "/foo";
const dirPath = "/foo/bar/";
const relativePath = relative(basePath, dirPath);
console.log(relativePath);
```

This will print `bar` in Node.js 22, but it will print `bar/` or `bar\` in Node.js 23 depending on the OS.

I believe that this change was intentional on Node.js' side, although looking at the public discussions it doesn't seems that they evaluated the effects on the ecosystem.

For `eslintrc` the new behavior is breaking some of the file enumeration logic that's been working for years, specifically in https://github.com/eslint/eslintrc/blob/98fb12d106f7f76132e64bce08aea4b79b45208f/lib/config-array/ignore-pattern.js#L91:

```js
const relPath = path.relative(from, to);
```

In the above line, when `to` ends with a path separator, the resulting `relPath` will also contain a trailing path separator in Node.js 23. This separator was not present in previous versions of Node.js, and it produces a relative path that will fail to work as intended in the code that follows. In the case of the failing unit test in ESLint, this results in a subdirectory being ignored unexpectedly.

If the legacy config system should continue to work in new version of Node.js then we need to fix the logic in `eslintrc` and update the dependency version in `eslint`. If not, then we should exclude tests for the legacy config from running in Node.js 23.

**Comment by mdjermanovic** [comment]
Looks like they are going to revert the change?

https://github.com/nodejs/node/pull/55414

**Comment by fasttime** [comment]
> Looks like they are going to revert the change?
> 
> [nodejs/node#55414](https://github.com/nodejs/node/pull/55414)

Hopefully yes, I didn't notice that PR. We should keep an eye on the next Node.js release to see if it fixes our issue.

**Comment by fasttime** [comment]
It looks like all test are passing again with [Node.js 23.1.0](https://github.com/nodejs/node/releases/tag/v23.1.0), which is now also picked by GitHub Actions' [setup-node](https://github.com/actions/setup-node) to run tests in CI.

* eslint/eslint#19055
* eslint/eslintrc#170

**Comment by fasttime** [comment]
This is solved now, so closing.
